src/parser/toplevel.lisp

```diff
@@ -554,56 +544,214 @@ consume all attributes"))))
+  (let ((iter (cons form file)))
+    (consume-symbol iter
+                    (lambda (symbol)
+                      (unless (string= (symbol-name symbol) "PACKAGE")
+                        (package-syntax-error (cst:first form) file "package declarations must start with `package`"))))
+    (let ((package (make-instance 'toplevel-package
+                     :name (symbol-name (consume-symbol iter))
+                     :docstring (consume-string iter :optional t))))
+      (loop :while (next-value-p iter)
+            :do (parse-package-clause package (consume-value iter) file))
+      package)))
 
-  ;; Package declarations must have a name
-  (unless (cst:consp (cst:rest form))
-    (error 'parse-error
-           :err (coalton-error
-                 :span (cst:source (cst:first form))
-                 :file file
-                 :message "Malformed package declaration"
-                 :primary-note "missing package name")))
-
-  ;; Package declarations cannot contain more than two forms
-  (when (cst:consp (cst:rest (cst:rest form)))
-    (error 'parse-error
-           :err (coalton-error
-                 :span (cst:source (cst:first (cst:rest (cst:rest form))))
-                 :file file
-                 :message "Malformed package declaration"
-                 :primary-note "unexpected forms")))
-
-  (unless (identifierp (cst:raw (cst:second form)))
-    (error 'parse-error
-           :err (coalton-error
-                 :span (cst:source (cst:second form))
-                 :file file
-                 :message "Malformed package declaration"
-                 :primary-note "package name must be a symbol")))
-
-  (let* ((package-name (symbol-name (cst:raw (cst:second form))))
-
-         (package (find-package package-name)))
-
-    (unless package
-      (setf package (make-package package-name :use '("COALTON" "COALTON-PRELUDE"))))
+;; Empty package for reading (package) forms
 
-    package))
+(defpackage #:coalton-impl/parser/%defpackage
+  (:documentation "The sole purpose of this package is to provide storage for symbols that are interned while Coalton package forms are read.")
+  (:use))
 
+(defun read-toplevel-package (stream file)
+  "Read and parse a Coalton toplevel package form."
+  (let ((*package* (find-package 'coalton-impl/parser/%defpackage)))
+    (unwind-protect
+         (multiple-value-bind (form presentp)
+             (maybe-read-form stream *coalton-eclector-client*)
+           (unless presentp
+             (error 'parse-error
+                    :err (coalton-error
+                          :span (cons (- (file-position stream) 2)
+                                      (- (file-position stream) 1))
+                          :file file
+                          :message "Unexpected EOF"
+                          :primary-note "missing package form")))
+           (parse-package form file))
+      (do-symbols (symbol)
+        (unintern symbol)))))
+
+(defun lisp-package (toplevel-package)
```

Comment:
This function and `make-defpackage` will have subtly different behavior. They need to share a single implementation.